api/s2n.h

```diff
@@ -687,6 +687,269 @@ extern int s2n_cert_get_utf8_string_from_extension_data_length(const uint8_t *ex
 S2N_API 
 extern int s2n_cert_get_utf8_string_from_extension_data(const uint8_t *extension_data, uint32_t extension_len, uint8_t *out_data, uint32_t *out_len);
 
+/* Pre-shared key (PSK) Hash Algorithm - RFC 8446 Section-2.2 */
+/* https://tools.ietf.org/html/rfc8446#section-2.2 */
+typedef enum {
+    S2N_PSK_HMAC_SHA256,
+    S2N_PSK_HMAC_SHA384,
+} s2n_psk_hmac;
+
+struct s2n_psk;
+
+/**
+ * Creates a new s2n external Pre-shared key (PSK) object with `S2N_PSK_HMAC_SHA256` as the default 
+ * PSK hash algorithm. An external PSK is an out-of-band provisioned PSK established 
+ * externally using a secure mutually agreed upon mechanism. 
+ * 
+ * # Safety 
+ *
+ * Use `s2n_psk_free` to free the memory allocated to the s2n external PSK object created by this API. 
+ *
+ * The external PSK authentication mechanism in >= TLS1.3 implicitly assumes each PSK is known to exactly one client 
+ * and one server, and these peers never switch roles. If this assumption is violated, the security properties of 
+ * >= TLS1.3 are severely weakened
+ *
+ * @return struct s2n_psk* Returns a pointer to the newly created external PSK object.
+ */
+S2N_API
+struct s2n_psk* s2n_external_psk_new();
+
+/**
+ * Frees the memory associated with the `s2n_psk` object.
+ *
+ * @param psk Pointer to the PSK object to be freed.
+ */
+S2N_API 
+int s2n_psk_free(struct s2n_psk **psk);
+
+/**
+ * Sets the PSK identity for a given s2n_psk object.
+ * The PSK identity is a unique identifier for the pre-shared secret.
+ * It is a non-secret value represented by raw bytes.
+ *
+ * # Safety 
+ *
+ * The PSK identity is transmitted over the network unencrypted and is a non-secret value.
+ * Do not include confidential information in the PSK identity.
+ * 
+ * Note that the identity is copied into s2n memory and the user is responsbile for freeing the memory 
+ * associated with the `identity` input. 
+ *
+ * @param psk A pointer to a PSK object to be updated with the identity.
+ * @param identity The identity in raw bytes format to be copied.
+ * @param identity_size The length of the PSK identity being set.
+ */
+S2N_API 
+int s2n_psk_set_identity(struct s2n_psk *psk, const uint8_t *identity, uint16_t identity_size);
+
+/**
+ * Sets the out-of-band/externally provisioned PSK secret for a given s2n_psk object.
+ *
+ * # Safety
+ *
+ * Note that the secret is copied into s2n memory and the user is responsbile for freeing the memory 
+ * associated with the `secret` input. 
+ *
+ * When using an out-of-band provisioned pre-shared secret, a critical consideration is using 
+ * sufficient entropy during the PSK generation. Deriving a shared secret from a password or 
+ * other low-entropy sources is not secure. A low-entropy secret, or password, is subject to 
+ * dictionary attacks based on the PSK binder. Refer: RFC8446 Section-2.2 for more information.
```

Comment:
Rather than copy/paste the warning from the RFC, can we try to simplify it?
```suggestion
 * WARNING: Deriving a shared secret from a password or other low-entropy source
 * is not secure and is subject to dictionary attacks.
 * See https://tools.ietf.org/rfc/rfc8446#section-2.2 for more information.
```